x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/ilm/ExecuteStepsUpdateTask.java

```diff
@@ -34,6 +39,7 @@
     private final PolicyStepsRegistry policyStepsRegistry;
     private final IndexLifecycleRunner lifecycleRunner;
     private final LongSupplier nowSupplier;
+    private final Map<String, Step.StepKey> indexToStepKeysForAsyncActions;
```

Comment:
I think this generally looks good, thanks for fixing this tricky one Lee! 

I'm only a bit worried for this being a `Map`. Namely, I'm still mulling over the implications of this not maintaining order and the fact that it can contain multiple indices. 

Multiple, consecutive `ClusterStateAction` steps could add to this `Map`. We don't have this situation **yet** but I'd like to think about what it would mean if we did. 

Would for now be ok to add an assertion that the size of this Map is `1` in `onClusterStateProcessed`?

Reply:
This is I think a side-effect of the use of `LegacyClusterTaskResultActionListener` in `IndexLifecycleRunner#ILM_TASK_EXECUTOR`. When that's addressed we will be able to pass data directly from the `doExecute()` method to the publish handler rather than relying on mutable fields on each task (that's exactly why https://github.com/elastic/elasticsearch/issues/83784 exists).

Reply:
> Multiple, consecutive ClusterStateAction steps could add to this Map. We don't have this situation yet but I'd like to think about what it would mean if we did.

This is part of why I used a map rather than a single `Tuple`. We don't have that situation now, but it would be handled in the event that we do in the future.

> Would for now be ok to add an assertion that the size of this Map is 1 in onClusterStateProcessed?

I don't think we need to enforce order here, as each index would be independent from the others, so I'm not sure whether we need to enforce it to always be 1? (Even though right now it's always going to be one, this will handle > 1 indices correctly)